File: platform/jvm/compiler/flabbergast/Parser.java
package flabbergast;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import flabbergast.Generator.ParameterisedBlock;

/**
 * The input being parsed along with all the memorised information from the
 * pack-rat parsing.
 */
public class Parser {
	/**
	 * A memory from previously-parsed passes and the final parse state.
	 */
	class Memory {
		int column;
		int index;
		AstNode result;
		int row;
	}

	/**
	 * The current position during parsing
	 */
	class Position {
		private int column;
		private ErrorCollector error_collector;
		private int index;
		private int row;

		private int trace_depth;

		Position(ErrorCollector error_collector) {
			index = 0;
			row = 1;
			column = 1;
			trace_depth = 0;
			this.error_collector = error_collector;
		}

		/**
		 * Create a new memory based on the type of the result.
		 * 
		 * @param result
		 *            The result to cache, or null if parsing failed.
		 */
		<T extends AstNode> void cache(int start_index, T result, Class<T> clazz) {
			if (!cache.containsKey(start_index)) {
				cache.put(start_index,
						new HashMap<Class<? extends AstNode>, Parser.Memory>());
			}
			Memory memory = new Memory();
			memory.result = result;
			memory.index = index;
			memory.row = row;
			memory.column = column;
			cache.get(start_index).put(clazz, memory);
		}

		/**
		 * Create a new memory based on the name of the result.
		 * 
		 * @param result
		 *            The result to cache, or null if parsing failed.
		 */
		public <T extends AstNode> void cache(String name, int start_index,
				T result) {
			if (!alternate_cache.containsKey(start_index)) {
				alternate_cache.put(start_index,
						new HashMap<String, Parser.Memory>());
			}
			Memory memory = new Memory();
			memory.result = result;
			memory.index = index;
			memory.row = row;
			memory.column = column;
			alternate_cache.get(start_index).put(name, memory);
		}

		/**
		 * Determine if the current position has been parsed based on the type
		 * of the rule.
		 */
		@SuppressWarnings("unchecked")
		<U extends AstNode, T extends U> boolean checkCache(Ptr<U> result,
				Class<T> clazz) {
			if (cache.containsKey(index) && cache.get(index).containsKey(clazz)) {
				Memory memory = cache.get(index).get(clazz);
				result.set((U) memory.result);
				index = memory.index;
				row = memory.row;
				column = memory.column;
				if (trace) {
					for (int it = 1; it < trace_depth; it++) {
						System.out.print(" ");
					}
					System.out.printf("%d:%d %s %s\n", row, column,
							(result.get() == null ? " M " : " H "),
							clazz.getName());
					trace_depth++;
				}
				return true;
			}
			return false;
		}

		/**
		 * Determine if the current position has been parsed based on the name
		 * of the rule.
		 */
		@SuppressWarnings("unchecked")
		<T extends AstNode> boolean checkCache(String name, Ptr<T> result) {
			if (alternate_cache.containsKey(index)
					&& alternate_cache.get(index).containsKey(name)) {
				Memory memory = alternate_cache.get(index).get(name);
				result.set((T) memory.result);
				index = memory.index;
				row = memory.row;
				column = memory.column;
				if (trace) {
					for (int it = 1; it < trace_depth; it++) {
						System.out.print(" ");
					}
					System.out.printf("%d:%d %s %s\n", row, column,
							(result.get() == null ? " M " : " H "), name);
					trace_depth++;
				}
				return true;
			}
			return false;
		}

		Position dup() {
			Position child = new Position(error_collector);
			child.index = index;
			child.row = row;
			child.column = column;
			return child;
		}

		int getColumn() {
			return column;
		}

		String getFileName() {
			return Parser.this.file_name;
		}

		int getIndex() {
			return index;
		}

		Parser getParser() {
			return Parser.this;
		}

		int getRow() {
			return row;
		}

		boolean isFinished() {
			return index >= input.length();
		}

		/**
		 * Match a fixed string in the input.
		 */
		boolean match(String word) {
			for (int it = 0; it < word.length(); it++) {
				if (word.charAt(it) != next()) {
					return false;
				}
			}
			return true;
		}

		void nameConstraint(String name) {
			error_collector.reportParseError(file_name, index, row, column,
					"The name " + name + " is already in use in this context.");
		}

		/**
		 * Consume a character from the input and return it.
		 */
		char next() {
			if (index < input.length()) {
				char c = input.charAt(index++);
				if (c == '\n') {
					row++;
					column = 0;
				} else {
					column++;
				}
				return c;
			} else {
				return '\0';
			}
		}

		/**
		 * Look back at the previously consumed character.
		 */
		char peekLast() {
			if (index > 0 && index <= input.length()) {
				return input.charAt(index - 1);
			} else {
				return '\0';
			}
		}

		/**
		 * Look the character to be consumed.
		 */
		char peekNext() {
			if (index > 0 && index < input.length()) {
				return input.charAt(index);
			} else {
				return '\0';
			}
		}

		/**
		 * Start a tracing block.
		 * 
		 * Must be matched to a finish.
		 */
		void traceEnter(String rule) {
			if (trace) {
				trace_depth++;
				for (int it = 1; it < trace_depth; it++) {
					System.out.print(" ");
				}
				System.out.printf("%d:%d > %s\n", row, column, rule);
			}
		}

		/**
		 * Finish a tracing block.
		 * 
		 * Must be matched to a start.
		 */
		void traceExit(String rule, boolean success) {
			if (trace) {
				trace_depth--;
				for (int it = 1; it < trace_depth; it++) {
					System.out.print(" ");
				}
				System.out.printf("%d:%s <%s %s\n", row, column, (success
						? "+"
						: "-"), rule);
			}
		}

		/**
		 * Mark an error at the current position.
		 */
		void update(String message, String syntax_name) {
			if (index > Parser.this.index + 1) {
				message = "Expected " + message + " while parsing "
						+ syntax_name + " but got "
						+ Parser.toLiteral("" + peekLast()) + " instead.";
				Parser.this.row = row;
				Parser.this.column = column;
				Parser.this.index = index;
			}
		}
	}

	private static boolean isWhiteSpace(String input) {
		for (int it = 0; it < input.length(); it++) {
			if (!Character.isWhitespace(input.charAt(it))) {
				return false;
			}
		}
		return true;
	}

	/**
	 * Open a file for parsing.
	 * 
	 * @throws IOException
	 */
	public static Parser open(String filename) throws IOException {
		return new Parser(filename, new String(Files.readAllBytes(Paths
				.get(filename))));
	}

	/**
	 * Format a string for public viewing.
	 */
	public static String toLiteral(String input) {
		if (input.length() == 0 || isWhiteSpace(input)) {
			return "whitespace";
		} else {
			return "one of \"" + input + "\"";
		}
	}

	Map<Integer, Map<String, Memory>> alternate_cache = new HashMap<Integer, Map<String, Memory>>();
	Map<Integer, Map<Class<? extends AstNode>, Memory>> cache = new HashMap<Integer, Map<Class<? extends AstNode>, Memory>>();
	int column;
	private String file_name;
	/**
	 * The position of the most helpful error yet produced by the parser.
	 */
	int index = -1;
	/**
	 * The characters to parse.
	 */
	String input;

	/**
	 * The most helpful error yet produced by the parser.
	 * 
	 * This may have a value even if the parse was successful.
	 */
	String message;

	int row;

	private boolean trace;

	public Parser(String filename, String input) {
		file_name = filename;
		this.input = input;
	}

	/**
	 * The name of the file being parsed for debugging and error information.
	 */
	public String getFileName() {
		return file_name;
	}

	/**
	 * Whether to produce copious junk on standard output.
	 */
	public boolean getTrace() {
		return trace;
	}

	/**
	 * Parse in the “file” context defined in the language specification.
	 */
	public <T> T parseFile(ErrorCollector collector, CompilationUnit<T> unit,
			String type_name) throws Exception {
		final Ptr<file> result = new Ptr<file>();
		Ptr<Position> position = new Ptr<Position>(new Position(collector));
		if (file.parseRule_Base(position, result)
				&& position.get().isFinished()) {
			if (result.get().analyse(collector)) {
				return unit.createRootGenerator(result.get(), type_name,
						new Generator.Block() {

							@Override
							public void invoke(final Generator generator)
									throws Exception {
								result.get()
										.generate(
												generator,
												new ParameterisedBlock<LoadableValue>() {

													@Override
													public void invoke(
															LoadableValue value)
															throws Exception {
														generator
																.doReturn(value);

													}

												});

							}
						});
			}
		} else {
			collector.reportParseError(file_name, index, row, column, message);
		}
		return null;
	}

	/**
	 * Parse a rule and, if successful, put the result into the provided list.
	 */
	<T> boolean parseIntoList(Ptr<Position> position, List<T> result,
			ParseRule<T> rule) {
		Ptr<T> obj = new Ptr<T>();
		if (rule.invoke(position, obj)) {
			result.add(obj.get());
			return true;
		} else {
			return false;
		}
	}

	/**
	 * Parse in the “repl” context defined in the language specification.
	 */
	public <T> T parseRepl(ErrorCollector collector, CompilationUnit<T> unit,
			String type_name) throws Exception {
		final Ptr<repl> result = new Ptr<repl>();
		Ptr<Position> position = new Ptr<Position>(new Position(collector));
		if (repl.parseRule_Base(position, result)
				&& position.get().isFinished()) {
			if (result.get().analyse(collector)) {
				return unit.createReplGenerator(result.get(), type_name,
						new ReplGenerator.Block() {

							@Override
							public void invoke(final Generator generator,
									LoadableValue root_frame,
									LoadableValue current_frame,
									LoadableValue update_current,
									LoadableValue escape_value,
									LoadableValue print_value

							) throws Exception {
								result.get()
										.generate(
												generator,
												root_frame,
												current_frame,
												update_current,
												escape_value,
												print_value,
												new ParameterisedBlock<LoadableValue>() {

													@Override
													public void invoke(
															LoadableValue value)
															throws Exception {
														generator
																.doReturn(value);

													}

												});

							}
						});
			}
		} else {
			collector.reportParseError(file_name, index, row, column, message);
		}
		return null;
	}

	public void setTrace(boolean trace) {
		this.trace = trace;
	}
}

File: platform/cli/parser.cs
using System;
using System.Collections.Generic;
using System.Xml;
namespace Flabbergast {

/**
 * Delegate used for parsing lists in a generic way from existing parse rules.
 */
internal delegate bool ParseRule<T>(ref ParserPosition position, out T result);

/**
 * The base class for nodes in the syntax tree.
 *
 * This is present to hold all the fields needed to locate the syntax element
 * in the original file for generating debugging and error information.
 */
internal abstract class AstNode : CodeRegion {
	public int StartRow { get; internal set; }
	public int StartColumn { get; internal set; }
	public int EndRow { get; internal set; }
	public int EndColumn { get; internal set; }
	public string FileName { get; internal set; }

	public abstract string PrettyName { get; }

	public abstract void GenerateApi(ApiGenerator api_gen, bool collect_names);
}
/**
 * The input being parsed along with all the memorised information from the pack-rat parsing.
 */
public class Parser {
	/**
	 * A memory from previously-parsed passes and the final parse state.
	 */
	internal struct Memory {
		internal Object Result;
		internal int Index;
		internal int Row;
		internal int Column;
	}
	internal Dictionary<int, Dictionary<System.Type, Memory>> Cache = new Dictionary<int, Dictionary<System.Type, Memory>>();
	internal Dictionary<int, Dictionary<string, Memory>> AlternateCache = new Dictionary<int, Dictionary<string, Memory>>();

	/**
	 * The name of the file being parsed for debugging and error information.
	 */
	public string FileName { get; private set; }
	/**
	 * The characters to parse.
	 */
	internal string Input;
	/**
	 * The position of the most helpful error yet produced by the parser.
	 */
	internal int Index = -1;
	/**
	 * The most helpful error yet produced by the parser.
	 *
	 * This may have a value even if the parse was successful.
	 */
	internal string Message;
	internal int Row;
	internal int Column;
	/**
	 * Whether to produce copious junk on standard output.
	 */
	public bool Trace { get; set; }

	/**
	 * Open a file for parsing.
	 */
	public static Parser Open(string filename) {
		var file = new System.IO.StreamReader(filename);
		var parser = new Parser(filename, file.ReadToEnd());
		file.Close();
		return parser;
	}

	/**
	 * Format a string for public viewing.
	 */
	public static string ToLiteral(string input) {
		if (String.IsNullOrWhiteSpace(input)) {
			return "whitespace";
		} else {
			return "one of \"" + input + "\"";
		}
	}

	public Parser(string filename, string input) {
		FileName = filename;
		Input = input;
	}
	public XmlDocument DocumentFile(ErrorCollector collector, string lib_name, string github) {
		file result;
		var position = new ParserPosition(this, collector);
		if (file.ParseRule_Base(ref position, out result) && position.Finished) {
			if (result.Analyse(collector)) {
				var api = ApiGenerator.Create(lib_name, github);
				result.GenerateApi(api, true);
				return api.Document;
			}
		} else {
			collector.ReportParseError(FileName, Index, Row, Column, Message);
		}
		return null;
	}
	/**
	 * Parse in the “file” context defined in the language specification.
	 */
	public System.Type ParseFile(ErrorCollector collector, CompilationUnit unit, string type_name) {
		file result;
		var position = new ParserPosition(this, collector);
		if (file.ParseRule_Base(ref position, out result) && position.Finished) {
			if (result.Analyse(collector)) {
				return unit.CreateRootGenerator(result, type_name, generator => result.Generate(generator, generator.Return));
			}
		} else {
			collector.ReportParseError(FileName, Index, Row, Column, Message);
		}
		return null;
	}
	/**
	 * Parse in the “repl” context defined in the language specification.
	 */
	public System.Type ParseRepl(ErrorCollector collector, CompilationUnit unit, string type_name) {
		repl result;
		var position = new ParserPosition(this, collector);
		if (repl.ParseRule_Base(ref position, out result) && position.Finished) {
			if (result.Analyse(collector)) {
				return unit.CreateReplGenerator(result, type_name,
					(generator, root, current, update_current, escape_value, print_value) =>
						result.Generate(generator, root, current, update_current, escape_value, print_value,
							 generator.Return));
			}
		} else {
			collector.ReportParseError(FileName, Index, Row, Column, Message);
		}
		return null;
	}
}
/**
 * The current position during parsing
 */
internal class ParserPosition {
	internal Parser Parser { get; private set; }
	internal int Index { get; private set; }
	internal int Row { get; private set; }
	internal int Column { get; private set; }
	internal bool Finished { get { return Index >= Parser.Input.Length; } }
	private readonly ErrorCollector error_collector;
	private int TraceDepth;
	internal ParserPosition(Parser parser, ErrorCollector error_collector) {
		Parser = parser;
		Index = 0;
		Row = 1;
		Column = 1;
		TraceDepth = 0;
		this.error_collector = error_collector;
	}
	/**
	 * Determine if the current position has been parsed based on the type of the rule.
	 */
	internal bool CheckCache<T, U>(out U result) where T: U {
		if (Parser.Cache.ContainsKey(Index) && Parser.Cache[Index].ContainsKey(typeof(T))) {
			var memory = Parser.Cache[Index][typeof(T)];
			result = (U) memory.Result;
			Index = memory.Index;
			Row = memory.Row;
			Column = memory.Column;
			if (Parser.Trace) {
				for (var it = 1; it < TraceDepth; it++) {
					Console.Write(" ");
				}
				Console.WriteLine(Row + ":" + Column + (result == null ? " M " : " H ") + typeof(T));
				TraceDepth++;
			}
			return true;
		}
		result = default(T);
		return false;
	}
	/**
	 * Determine if the current position has been parsed based on the name of the rule.
	 */
	internal bool CheckCache<T>(string name, out T result) {
		if (Parser.AlternateCache.ContainsKey(Index) && Parser.AlternateCache[Index].ContainsKey(name)) {
			var memory = Parser.AlternateCache[Index][name];
			result = (T) memory.Result;
			Index = memory.Index;
			Row = memory.Row;
			Column = memory.Column;
			if (Parser.Trace) {
				for (var it = 1; it < TraceDepth; it++) {
					Console.Write(" ");
				}
				Console.WriteLine(Row + ":" + Column + (result == null ? " M " : " H ") + name);
				TraceDepth++;
			}
			return true;
		}
		result = default(T);
		return false;
	}
	/**
	 * Create a new memory based on the type of the result.
	 *
	 * <param name="result">The result to cache, or null if parsing failed.</param>
	 */
	internal void Cache<T>(int start_index, T result) {
		if (!Parser.Cache.ContainsKey(start_index)) {
			Parser.Cache[start_index] = new Dictionary<System.Type, Parser.Memory>();
		}
		Parser.Cache[start_index][typeof(T)] = new Parser.Memory() { Result = result, Index = Index, Row = Row, Column = Column };
	}
	/**
	 * Create a new memory based on the name of the result.
	 *
	 * <param name="result">The result to cache, or null if parsing failed.</param>
	 */
	public void Cache<T>(string name, int start_index, T result) {
		if (!Parser.AlternateCache.ContainsKey(start_index)) {
			Parser.AlternateCache[start_index] = new Dictionary<string, Parser.Memory>();
		}
		Parser.AlternateCache[start_index][name] = new Parser.Memory() { Result = result, Index = Index, Row = Row, Column = Column };
	}
	internal ParserPosition Clone() {
	    var child = new ParserPosition(Parser, error_collector)
	    {
	        Index = Index,
	        Row = Row,
	        Column = Column,
	        TraceDepth = TraceDepth
	    };
	    return child;
	}
	/**
	 * Consume a character from the input and return it.
	 */
	internal Char Next() {
		if (Index < Parser.Input.Length) {
			var c = Parser.Input[Index++];
			if (c == '\n') {
				Row++;
				Column = 0;
			} else {
				Column++;
			}
			return c;
		} else {
			return '\0';
		}
	}
	/**
	 * Look back at the previously consumed character.
	 */
	internal Char PeekLast() {
		if (Index > 0 && Index <= Parser.Input.Length) {
			return Parser.Input[Index - 1];
		} else {
			return '\0';
		}		
	}
	/**
	 * Match a fixed string in the input.
	 */
	internal bool Match(string word) {
		for (var it = 0; it < word.Length; it++) {
			if (word[it] != Next()) {
				return false;
			}
		}
		return true;
	}
	/**
	 * Parse a rule and, if succcessful, put the result into the provided list.
	 */
	internal static bool ParseIntoList<T>(ref ParserPosition position, List<T> result, ParseRule<T> rule) {
		T obj;
		if (rule(ref position, out obj)) {
			result.Add(obj);
			return true;
		} else {
			return false;
		}
	}
	/**
	 * Start a tracing block.
	 *
	 * Must be matched to a finish.
	 */
	internal void TraceEnter(string rule) {
		if (Parser.Trace) {
			TraceDepth++;
			for (var it = 1; it < TraceDepth; it++) {
				Console.Write(" ");
			}
			Console.WriteLine(Row + ":" + Column + " > " + rule);
		}
	}

	/**
	 * Finish a tracing block.
	 *
	 * Must be matched to a start.
	 */
	internal void TraceExit(string rule, bool success) {
		if (Parser.Trace) {
			TraceDepth--;
			for (var it = 1; it < TraceDepth; it++) {
				Console.Write(" ");
			}
			Console.WriteLine(Row + ":" + Column + " <" + (success ? "+" : "-") + " " + rule);
		}
	}

	/**
	 * Mark an error at the current position.
	 */
	internal void Update(string message, string syntax_name) {
		if (Index > Parser.Index) {
			Parser.Message = "Expected " + message + " while parsing " + syntax_name + " but got " + Parser.ToLiteral(PeekLast().ToString()) + " instead.";
			Parser.Row = Row;
			Parser.Column = Column;
			Parser.Index = Index;
		}
	}

	internal void NameConstraint(string name) {
		error_collector.ReportParseError(Parser.FileName, Index, Row, Column, "The name " + name + " is already in use in this context.");
	}
}
}
